src/lib/threadpool.c

```diff
@@ -0,0 +1,368 @@
+#include "src/lib/threadpool.h"
+#include <assert.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <string.h>
+#include <uv.h>
+#include "src/utils.h"
+
+static struct xx_loop_s *xx_loop(struct uv_loop_s *loop)
+{
+	return (struct xx_loop_s *)loop;
+}
+
+#define xx__has_active_reqs(loop) ((loop)->active_reqs > 0)
+
+#define xx__req_register(loop, req)    \
+	do {                           \
+		(loop)->active_reqs++; \
+	} while (0)
+
+#define xx__req_unregister(loop, req)              \
+	do {                                       \
+		assert(xx__has_active_reqs(loop)); \
+		(loop)->active_reqs--;             \
+	} while (0)
+
+enum {
+	MAX_THREADPOOL_SIZE = 1024,
+	XX_THREADPOOL_SIZE = 4,
+	NOT_AFFINED = ~0u,
+};
+
+struct thread_args
+{
+	uv_sem_t *sem;
+	unsigned int idx;
+};
+
+static uv_cond_t cond;
+static uv_mutex_t mutex;
+static unsigned int idle_threads;
+static unsigned int nthreads;
+static uv_thread_t *threads;
+static struct thread_args *thread_args;
+static uv_key_t thread_key;
+static queue *thread_queues;
+static queue exit_message;
+static queue wq;
+
+static void xx__cancelled(struct xx__work *w UNUSED)
+{
+	abort();
+}
+
+unsigned int xx__thread_id(void)
+{
+	return *(unsigned int *)uv_key_get(&thread_key);
+}
+
+/* To avoid deadlock with uv_cancel() it's crucial that the worker
+ * never holds the global mutex and the loop-local mutex at the same time.
+ */
+static void worker(void *arg)
+{
+	struct xx__work *w;
+	queue *q;
+	struct thread_args *ta = arg;
+
+	uv_key_set(&thread_key, &ta->idx);
+	uv_sem_post(ta->sem);
+	uv_mutex_lock(&mutex);
+	for (;;) {
+		/* `mutex` should always be locked at this point. */
+		/* Keep waiting while either no work is present */
+		while (QUEUE__IS_EMPTY(&wq) &&
+		       QUEUE__IS_EMPTY(&thread_queues[ta->idx])) {
+			idle_threads += 1;
+			uv_cond_wait(&cond, &mutex);
+			idle_threads -= 1;
+		}
+
+		/* Process work item thread affinity */
+		if (!QUEUE__IS_EMPTY(&thread_queues[ta->idx])) {
+			q = QUEUE__HEAD(&thread_queues[ta->idx]);
+			QUEUE__REMOVE(q);
+			QUEUE__INIT(q);
+			QUEUE__INSERT_HEAD(&wq, q);
+		}
```

Comment:
If we switch to having a separate input broadcast channel for checkpoint notifications from the main thread then we'll have to check that before we take an item off the thread-specific work queue, and switch to a different processing loop if a notification is found until we've executed the checkpoint. I'm not sure exactly how to integrate that here though